jwst/associations/lib/rules_level2_base.py

```diff
@@ -1027,3 +1030,147 @@ def _init_hook(self, item):
+                    min_diff = diff
+                    closest = direct
+
+        # Note the selected direct image. Used in `Asn_Lv2WFSS._get_opt_element`
+        self.direct_image = closest
+
+        # Remove all direct images from the association.
+        members = self.current_product['members']
+        direct_idxs = [
+            idx
+            for idx, member in enumerate(members)
+            if member['exptype'] == 'direct_image'
+        ]
+        deque((
+            list.pop(members, idx)
+            for idx in sorted(direct_idxs, reverse=True)
+        ))
+
+        # Add the Level3 catalog, direct image, and segmentation map members
+        lv3_direct_image_root = DMS_Level3_Base._dms_product_name(self)
+        members.append(
+            Member({
+                'expname': lv3_direct_image_root + '_i2d.fits',
+                'exptype': 'direct_image'
+            })
+        )
+        members.append(
+            Member({
+                'expname': lv3_direct_image_root + '_cat.ecsv',
+                'exptype': 'sourcecat'
+            })
+        )
+        members.append(
+            Member({
+                'expname': lv3_direct_image_root + '_segm.fits',
+                'exptype': 'segmap'
+            })
+        )
+
+    def finalize(self):
+        """Finalize the association
+
+        For WFSS, this involves taking all the direct image exposures,
+        determine which one is first after last science exposure,
+        and creating the catalog name from that image.
+        """
+        try:
+            self.add_catalog_members()
+        except AssociationNotValidError as err:
+            logger.debug(
+                '%s: %s',
+                self.__class__.__name__, str(err)
+            )
+            return None
+
+        return super(AsnMixin_Lv2WFSS, self).finalize()
+
+    def get_exposure_type(self, item, default='science'):
+        """Modify exposure type depending on dither pointing index
+
+        If an imaging exposure as been found, treat is as a direct image.
+        """
+        exp_type = super(AsnMixin_Lv2WFSS, self).get_exposure_type(
+            item, default
+        )
+        if exp_type == 'science' and item['exp_type'] in ['nis_image', 'nrc_image']:
+            exp_type = 'direct_image'
+
+        return exp_type
+
+    def _get_opt_element(self):
+        """Get string representation of the optical elements
+
+        Returns
+        -------
+        opt_elem: str
+            The Level3 Product name representation
+            of the optical elements.
+        Notes
+        -----
+        This is an override for the method in `DMSBaseMixin`.
+        The optical element is retrieved from the chosen direct image
+        found in `self.direct_image`, determined in the `self.finalize`
+        method.
+        """
+        item = self.direct_image.item
+        opt_elems = []
+        for keys in [['filter', 'band'], ['pupil', 'grating']]:
```

Comment:
I see `grating` referenced here in association (so to speak) with `pupil`. Keep in mind that in the NIRISS instrument the grisms used for WFSS are in the filter wheel and the blocking filters that go with them are in the pupil wheel (the opposite of NIRCam). So I hope that's accounted for here.

Reply:
So how and where is this now fixed?  I still see `pupil` being loaded as a constraint for NIS (below). Is that being used for the filter value, in the same way `filter` is used for the filter value in NRC?  I'm confused.

Reply:
Yes. NIRISS essentially has two filter wheels, one of which has some pupil-like spots. Hence the one wheel is called "pupil". The "filter" wheel has the grisms in it. Basically all mixed up. [See figure 3 here](https://jwst-docs.stsci.edu/jwst-near-infrared-imager-and-slitless-spectrograph#JWSTNearInfraredImagerandSlitlessSpectrograph-Opticalelements).

NIRCam was much cleaner, putting the grisms in their "pupil" wheel and all filters are in the "filter" wheel.

😞